include/ppx/metrics.h

```diff
@@ -195,58 +177,77 @@ class Run final
 
 public:
     template <typename T>
-    T* AddMetric(MetricMetadata metadata);
+    T* AddMetric(MetricMetadata metadata)
+    {
+        PPX_ASSERT_MSG(!metadata.name.empty(), "The metric name must not be empty");
+        PPX_ASSERT_MSG(!HasMetric(metadata.name), "Metrics must have unique names (duplicate name detected)");
+
+        auto* pMetric = new T(metadata);
+        auto  metric  = std::unique_ptr<T>(pMetric);
```

Comment:
The goal of an `unique_ptr` is to enforce this pointer is unique. Returning a raw pointer after sound like adding constraints internally while keeping the unsafe aspect.

Reply:
I'd be fine to move raw pointers to shared_ptr (same as the knob framwork), but not this half-mesure.

Reply:
Unique_ptr is more about ownership than it is about true uniqueness. Returning a raw pointer like this generally indicates that the caller does not OWN the memory in question, and that memory management is handled elsewhere; therein lies the uniqueness. Shared ptr has some very gnarly consequences and is dangerous for a variety of reasons; it should be avoided as much as possible. I will reach out to you specifically about this, because it's surprising that you would recommend it, and it seems like there's some misunderstanding of the value of unique_ptr.

Reply:
Yeah, you are right. It's just emplaced in owned storage, and the manager lifetime is currently bounded to the external references (`mMetrics` in `application.h`).

I was worried about dangling pointers, as yes, `shared_ptr`/`weak_ptr` does means atomicity overhead, but:
 - is this really a hot path? Doesn't seem like the overhead would be significant vs the compute/graphics load.
 - it allows us to actually catch dangling pointers. Which seems to be the main benefit to have an easy-to-use/safe API for users. (We don't have `raw_ptr` construct like in chrome, hence the suggestion to use shared+weak and not export the raw pointer)
 (We can talk about than in half an hour)

Reply:
Per discussion - we think the current solution is likely inadequate depending on the situation where this gets used. In particular, we have some use cases where this pointer might reach a bit beyond the application itself, and that leads to difficulty where you can't expire the pointer when you need to (without some separate signalling, etc.)

For now this is still an improvement over what was previously here, so I'll put in a few TODOs mentioning the situation, then clean it up as our use case becomes more mature.

I suspect long-term, we may want to avoid pointers altogether. The MANAGER is guaranteed to have the same lifetime as the application, but everything else has limits. Something like a LUT or a map with function calls may be more valuable, but all of this hinges on whether the access to these components remains single-threaded.